File: src/main/java/no/novari/flyt/configuration/mapping/ValuesFromCollectionMappingMappingService.java
package no.novari.flyt.configuration.mapping;

import no.novari.flyt.configuration.model.configuration.dtos.FromCollectionMappingDto;
import no.novari.flyt.configuration.model.configuration.dtos.ValueMappingDto;
import no.novari.flyt.configuration.model.configuration.entities.collection.ValuesFromCollectionMapping;
import org.springframework.stereotype.Service;


@Service
public class ValuesFromCollectionMappingMappingService {

    private final InstanceCollectionReferencesMappingService instanceCollectionReferencesMappingService;
    private final ValueMappingMappingService valueMappingMappingService;

    public ValuesFromCollectionMappingMappingService(
            InstanceCollectionReferencesMappingService instanceCollectionReferencesMappingService,
            ValueMappingMappingService valueMappingMappingService
    ) {
        this.instanceCollectionReferencesMappingService = instanceCollectionReferencesMappingService;
        this.valueMappingMappingService = valueMappingMappingService;
    }

    public ValuesFromCollectionMapping toEntity(
            FromCollectionMappingDto<ValueMappingDto> valuesFromCollectionMappingDto
    ) {
        return ValuesFromCollectionMapping
                .builder()
                .instanceCollectionReferencesOrdered(
                        instanceCollectionReferencesMappingService.toEntity(
                                valuesFromCollectionMappingDto.getInstanceCollectionReferencesOrdered()
                        )
                )
                .valueMapping(
                        valueMappingMappingService.toEntity(valuesFromCollectionMappingDto.getElementMapping())
                )
                .build();
    }

    public FromCollectionMappingDto<ValueMappingDto> toDto(ValuesFromCollectionMapping valuesFromCollectionMapping) {
        return FromCollectionMappingDto
                .<ValueMappingDto>builder()
                .instanceCollectionReferencesOrdered(
                        instanceCollectionReferencesMappingService.toDto(
                                valuesFromCollectionMapping.getInstanceCollectionReferencesOrdered()
                        )
                )
                .elementMapping(valueMappingMappingService.toDto(valuesFromCollectionMapping.getValueMapping()))
                .build();
    }

}

File: src/main/kotlin/no/novari/flyt/configuration/mapping/ValuesFromCollectionMappingMappingService.kt
package no.novari.flyt.configuration.mapping

import no.novari.flyt.configuration.model.configuration.dtos.FromCollectionMappingDto
import no.novari.flyt.configuration.model.configuration.dtos.ValueMappingDto
import no.novari.flyt.configuration.model.configuration.entities.collection.ValuesFromCollectionMapping
import org.springframework.stereotype.Service

@Service
class ValuesFromCollectionMappingMappingService(
    private val instanceCollectionReferencesMappingService: InstanceCollectionReferencesMappingService,
    private val valueMappingMappingService: ValueMappingMappingService,
) {
    fun toEntity(
        valuesFromCollectionMappingDto: FromCollectionMappingDto<ValueMappingDto>,
    ): ValuesFromCollectionMapping =
        ValuesFromCollectionMapping
            .builder()
            .instanceCollectionReferencesOrdered(
                instanceCollectionReferencesMappingService.toEntity(
                    valuesFromCollectionMappingDto.instanceCollectionReferencesOrdered,
                ),
            ).valueMapping(
                valueMappingMappingService.toEntity(
                    requireNotNull(valuesFromCollectionMappingDto.elementMapping),
                ),
            ).build()

    fun toDto(valuesFromCollectionMapping: ValuesFromCollectionMapping): FromCollectionMappingDto<ValueMappingDto> =
        FromCollectionMappingDto
            .builder<ValueMappingDto>()
            .instanceCollectionReferencesOrdered(
                instanceCollectionReferencesMappingService.toDto(
                    valuesFromCollectionMapping.instanceCollectionReferencesOrdered,
                ),
            ).elementMapping(
                valueMappingMappingService.toDto(
                    requireNotNull(valuesFromCollectionMapping.valueMapping),
                ),
            ).build()
}
